plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java

```diff
@@ -0,0 +1,307 @@
+import org.elasticsearch.action.ShardOperationFailedException;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.delete.DeleteRequest;
+import org.elasticsearch.action.delete.DeleteResponse;
+import org.elasticsearch.action.search.*;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHitField;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class TransportDeleteByQueryAction extends HandledTransportAction<DeleteByQueryRequest, DeleteByQueryResponse> {
+
+    private final TransportSearchAction searchAction;
+    private final TransportSearchScrollAction scrollAction;
+    private final Client client;
+
+    @Inject
+    protected TransportDeleteByQueryAction(Settings settings, ThreadPool threadPool, Client client,
+                                           TransportSearchAction transportSearchAction,
+                                           TransportSearchScrollAction transportSearchScrollAction,
+                                           TransportService transportService, ActionFilters actionFilters) {
+        super(settings, DeleteByQueryAction.NAME, threadPool, transportService, actionFilters, DeleteByQueryRequest.class);
+        this.searchAction = transportSearchAction;
+        this.scrollAction = transportSearchScrollAction;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+        new AsyncDeleteByQueryAction(request, listener).start();
+    }
+
+    class AsyncDeleteByQueryAction {
+
+        private final DeleteByQueryRequest request;
+        private final ActionListener<DeleteByQueryResponse> listener;
+
+        private final long startTime;
+        private final AtomicBoolean timedOut;
+        private final AtomicLong total;
+
+        private volatile ShardOperationFailedException[] shardFailures;
+        private final Map<String, IndexDeleteByQueryResponse> results;
+
+        AsyncDeleteByQueryAction(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+            this.startTime = System.currentTimeMillis();
```

Comment:
I didn't know about this class but for what I saw it captures the system current time at regular intervals, trading off a less time precision for better performance, right?

If so, I think we can use it here since millisecond precision for this potentially long process is not really necessary.